### PR TITLE
Make outgoing PUT error calls when handling errors

### PR DESCRIPTION
**Pull request by kkzeng**
Need to update: https://github.com/mojaloop/pisp/blob/master/docs/error_codes.md

### CODE REVIEW

File: src/domain/consents/generateChallenge.ts [hunk 1]
```diff
@@ -58,7 +59,12 @@ export async function updateConsentCredential (
 
   // Update in database,
   // relying on database validation for any null or relational aspects.
-  await consentDB.update(consent)
+  try {
+    await consentDB.update(consent)
+  } catch (error) {
+    throw new DatabaseError(consent.id)
```
Comment (eoln) [inline src/domain/consents/generateChallenge.ts:65]: you should log `error` here elsewhere we will know nothing about its nature

File: src/domain/consents/revoke.ts [hunk 2]
```diff
@@ -53,13 +54,18 @@ export async function revokeConsentStatus (
...
   }
 
   consent.status = 'REVOKED'
   consent.revokedAt = (new Date()).toISOString()
-  await consentDB.update(consent)
+
+  try {
+    await consentDB.update(consent)
+  } catch (error) {
+    throw new DatabaseError(consent.id)
```
Comment (eoln) [inline src/domain/consents/revoke.ts:66]: please log `error` here

File: src/server/handlers/consents/{ID}.ts [hunk 3]
```diff
@@ -70,13 +72,24 @@ export async function validateAndUpdateConsent (
...
     const consent: Consent = await retrieveValidConsent(consentId, challenge)
-    const verifyResult = verifySignature(challenge, signature, publicKey)
+    let verifyResult: boolean
+
+    // Use a nested try-catch to convert verifySignature errors to mojaloop
+    // accepted errors
+    try {
+      verifyResult = verifySignature(challenge, signature, publicKey)
+    } catch (error) {
+      // Signature validity was not determined
```
Comment (eoln) [inline src/server/handlers/consents/{ID}.ts:86]: log `error` here

File: src/server/handlers/consents/{ID}/generateChallenge.ts [hunk 4]
```diff
@@ -97,17 +99,21 @@ export async function generateChallengeAndPutConsent (
...
+      throw new ActiveConsentChallengeRequestError(id)
     }
 
     // Retrieve Scopes
-    const scopesRetrieved: Scope[] = await scopeDB.retrieveAll(id)
+    let scopesRetrieved: Scope[]
+    try {
+      scopesRetrieved = await scopeDB.retrieveAll(id)
+    } catch (error) {
+      // Convert error to Mojaloop understood error
```
Comment (eoln) [inline src/server/handlers/consents/{ID}/generateChallenge.ts:111]: log `error` here

**Review by eoln — CHANGES_REQUESTED** [review]
if you catch(error) and you throw an error of another type without inspecting the nature of `error` at least you should log it so later it will be easier to debug

File: src/domain/consents/generateChallenge.ts (hunk 1 above)
Comment (kkzeng) [reply src/domain/consents/generateChallenge.ts:65]: agreed

File: src/domain/errors.ts [hunk 5]
```diff
@@ -28,9 +28,9 @@ import { TErrorInformation, TErrorInformationObject } from '@mojaloop/sdk-standa
 import { thirdPartyRequest } from '../lib/requests'
 
 // Type guard for Mojaloop errors
-export function isMojaloopError(error: any): error is TErrorInformation {
+export function isMojaloopError (error: unknown): error is TErrorInformation {
   const mojaloopError = (error as TErrorInformation)
-  if(mojaloopError.errorCode && mojaloopError.errorDescription) {
+  if (mojaloopError.errorCode && mojaloopError.errorDescription) {
```
Comment (eoln) [inline src/domain/errors.ts:33]:
```suggestion
  return !!(mojaloopError.errorCode && mojaloopError.errorDescription)
```

**Review by eoln — CHANGES_REQUESTED** [review]

File: jest.bdd.config.js [hunk 6]
```diff
@@ -15,7 +15,7 @@ module.exports = {
     global: {
       statements: 20,
       functions: 20,
-      branches: 7,
+      branches: 5,
```
[the remark this answers is not in the document]
Comment (kkzeng) [reply jest.bdd.config.js:18]: I think we don't have bdd test in auth-service right now, they need to be written

File: src/domain/errors.ts [hunk 7]
```diff
@@ -19,42 +19,263 @@
...
+  try {
+    await
+    thirdPartyRequest.putThirdpartyRequestsTransactionsAuthorizationsError(
+      errorInfoObj,
+      consentId,
+      destParticipantId
+    )
+  } catch (error) {
+    Logger.push(error)
+    Logger.error('Could not make PUT error request')
```
[the remark this answers is not in the document]
Comment (kkzeng) [reply src/domain/errors.ts]: Agreed

File: src/domain/errors.ts [hunk 8]
```diff
@@ -19,42 +19,263 @@
...
+  public readonly errorDescription: string
 
   public constructor (consentId: string) {
     super(`Credential Payload not provided for ${consentId}`)
+    this.errorDescription = this.message
+    this.consentId = consentId
+  }
+}
+
+export class InvalidSignatureError extends Error implements TErrorInformation {
```
[the remark this answers is not in the document]
Comment (kkzeng) [reply src/domain/errors.ts]: I refactored to use inheritance

File: test/unit/domain/consents/revoke.test.ts [hunk 9]
```diff
@@ -119,7 +119,7 @@ describe('server/domain/consents/revoke', (): void => {
           .toThrowError(new DatabaseError(partialConsentActiveConflictingInitiatorId.id))
 
         expect(mockConsentUpdate).toHaveBeenCalled()
-        expect(mockLoggerPush).toHaveBeenCalledWith(testError)
```
Comment (eoln) [inline test/unit/domain/consents/revoke.test.ts]: that is strange - we are calling`push` in revoke.ts L82

**Review by eoln — CHANGES_REQUESTED** [review]

File: test/unit/domain/consents/revoke.test.ts (hunk 9 above)
Comment (kkzeng) [reply test/unit/domain/consents/revoke.test.ts]: I think you reviewed an old version. It says "Outdated" on the diff. 

Either way, it has been converted to ` expect(mocked(logger.push)).toHaveBeenCalledWith({ consent: expect.objectContaining(partialConsentActiveConflictingInitiatorId) })` so we are indeed expecting a push. Just using the newer syntax that you introduced in your logger PR.

File: src/domain/consents/generateChallenge.ts [hunk 10]
```diff
@@ -58,7 +59,14 @@ export async function updateConsentCredential (
 
   // Update in database,
   // relying on database validation for any null or relational aspects.
-  await consentDB.update(consent)
+  try {
+    await consentDB.update(consent)
+  } catch (error) {
+    Logger.push(error)
```
Comment (eoln) [inline src/domain/consents/generateChallenge.ts]: you can't use the Logger form central-services-logger

Comment (eoln) [reply src/domain/consents/generateChallenge.ts]: you need to chain `push` with `error`

File: src/domain/errors.ts [hunk 11]
```diff
@@ -19,42 +19,263 @@
...
+
+  try {
+    await
+    thirdPartyRequest.putThirdpartyRequestsTransactionsAuthorizationsError(
+      errorInfoObj,
+      consentId,
+      destParticipantId
+    )
+  } catch (error) {
+    Logger.push(error)
```
Comment (eoln) [inline src/domain/errors.ts]: change the logger

**Review by eoln — APPROVED** [review]
lgtm

**Pull request merged**